### PR TITLE
Experimental support for `otel` via `Backend`

### PR DESCRIPTION
**Pull request by am-kantox**
The new behaviour `Backend` has been introduced, making this library agnostic from the actual telemetry implementation.

@Hajto I understand it’s 811 LoCs. There are two examples with tests added, that’s mostly why.

What I have done:

- extracted the actual backend into a behaviour, allowing to transparently switch between `:telemetry`, `:opentelemetry` and whatnot implementing `Telemetria.Backend`
- the `OpenTelemetry` implementation is actually 20 LoCs
- introduced two example applications, one using `:telemetry`, another one using `:opentelemetry`, with the literally same source code
- made sure both works the same

I also dropped some comments into PR.

### CODE REVIEW

File: examples/otel/test/otel_test.exs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+    log =
+      capture_log(fn ->
+        assert Otel.f_to_c(451) == 233
+        Process.sleep(100)
+      end)
+
+    assert log =~ "[warning] Unexpected throttle setting for group `:weather_reports` → nil"
+
+    assert_receive {:span,
```
Comment (am-kantox) [inline examples/otel/test/otel_test.exs:24]: This is validating attributes _and_ events for `OpenTelemetry`.

File: examples/otel/test/support/with_span.ex [hunk 2]
```diff
@@ -0,0 +1,32 @@
+defmodule Otel.Test.WithSpan do
```
Comment (am-kantox) [inline examples/otel/test/support/with_span.ex:1]: Disregard, it’s a helper to tweak Otel outcome in `iex`.

File: examples/tm/test/tm_test.exs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  test "converts fahrenheit to celsius (capture)" do
+    log =
+      capture_log(fn ->
+        assert Tm.f_to_c(451) == 233
+        Process.sleep(100)
+      end)
+
+    assert log =~ "[warning] Unexpected throttle setting for group `:weather_reports` → nil"
+
+    assert log =~
```
Comment (am-kantox) [inline examples/tm/test/tm_test.exs:20]: Test for `:telemetry` backend with `console` instrumenter.

File: lib/telemetria.ex [hunk 4]
```diff
@@ -335,25 +344,27 @@ defmodule Telemetria do
             utc: DateTime.utc_now()
           ]
 
+          block_ctx = Backend.entry(unquote(event))
```
Comment (am-kantox) [inline lib/telemetria.ex:347]: First callback, we are entering the span.

File: lib/telemetria.ex [hunk 5]
```diff
@@ -335,25 +344,27 @@ defmodule Telemetria do
...
-            })
+            attributes = %{
+              env: unquote(caller),
+              locals: Keyword.take(binding(), unquote(locals)),
+              result: unquote(result_transform).(result),
+              args: unquote(args_transform).(unquote(args)),
+              context: unquote(context)
+            }
+
+            Backend.update(unquote(event), %{timestamp: now[:utc]})
```
Comment (am-kantox) [inline lib/telemetria.ex:362]: Adding some events. This should be re-thought and maybe amended with somewhat more useful.

File: lib/telemetria/backend/open_telemetry.ex [hunk 6]
```diff
@@ -0,0 +1,52 @@
+case {Telemetria.Application.open_telemetry?(), Code.ensure_compiled(OpenTelemetry)} do
+  {true, {:module, OpenTelemetry}} ->
+    defmodule Telemetria.Backend.OpenTelemetry do
```
Comment (am-kantox) [inline lib/telemetria/backend/open_telemetry.ex:3]: Implementation for `OpenTelemetry`. This is the most valuable part of this PR.

File: examples/otel/lib/otel/application.ex [hunk 7]
```diff
@@ -0,0 +1,20 @@
+defmodule Otel.Application do
+  # See https://hexdocs.pm/elixir/Application.html
+  # for more information on OTP Applications
+  @moduledoc false
+
+  use Application
+
+  @impl true
+  def start(_type, _args) do
```
Comment (Hajto) [inline examples/otel/lib/otel/application.ex:9]: Don't we need to start an exporter here?

File: test/telemetria_test.exs [hunk 8]
```diff
@@ -97,7 +97,7 @@ defmodule Telemetria.Test do
         assert 0 == Example.annotated_1(nil)
         assert :forty_two == Example.annotated_1("42")
         assert 42 == Example.annotated_1(42)
-        Process.sleep(500)
+        Process.sleep(100)
```
Comment (Hajto) [inline test/telemetria_test.exs:100]: Documenting why this change happened in this pr my be helpful

**Comment by Hajto** [comment]
It looks very promising can I take it for a spin in one of our testing environments to see how well it works in that environment?

File: test/telemetria_test.exs (hunk 8 above)
Comment (am-kantox) [reply test/telemetria_test.exs:100]: To pass the test suite faster.

**Comment by am-kantox** [comment]
> It looks very promising can I take it for a spin in one of our testing environments to see how well it works in that environment?

What do you mean “can”? It’ s public domain.

I would better let me merge and release it first, though. Could happen in a matter of an hour.
